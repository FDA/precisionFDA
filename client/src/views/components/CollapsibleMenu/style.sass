@import "../../../styles/common.sass"
@import "../../../styles/variables.sass"

.collapsible-menu
  .title
    @extend .pfda-subsection-heading
    display: inline-block
    letter-spacing: 0

    a
      color: $color-text-medium-grey

      &:hover
        color: $color-text-dark-grey

.collapsible-header
  .title
    @extend .pfda-subsection-heading
    letter-spacing: 0

    &:hover
      color: $color-text-dark-grey
      cursor: pointer

  #toggle-icon
    cursor: pointer
    margin-right: 6px
    transition: all 0.5s;
    -webkit-transform: rotate(0deg)
    -moz-transform: rotate(0deg)
    transform: rotate(0deg)

    &.collapsed
      -webkit-transform: rotate(180deg)
      -moz-transform: rotate(180deg)
      transform: rotate(180deg)

.collapsible-menu__item
  display: block
  padding-left: 16px

  &--disabled
    color: #888

  a
    color: $color-text-medium-grey

    &:hover
      color: $color-text-dark-grey
