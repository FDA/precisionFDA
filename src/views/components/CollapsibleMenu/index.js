import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { HashLink } from 'react-router-hash-link'
import styled from 'styled-components'
import Icon from '../Icon'
import { colors } from '../../../styles/theme'

const StyledMenu = styled.div`
.collapsible-menu {
  .title {
    display: inline-block;
    letter-spacing: 0;

    a {
      color: ${colors.textMediumGrey};

      &:hover {
        color: ${colors.textDarkGrey};
      }
    }
  }
}
.collapsible-header {
  .title {
    letter-spacing: 0;

    &:hover {
      color: ${colors.textDarkGrey};
      cursor: pointer;
    }
  }
  #toggle-icon {
    cursor: pointer;
    margin-right: 6px;
    transition: all 0.5s;

    -webkit-transform: rotate(0deg);
    -moz-transform: rotate(0deg);
    transform: rotate(0deg);

    &.collapsed {
      -webkit-transform: rotate(180deg);
      -moz-transform: rotate(180deg);
      transform: rotate(180deg);
    }
  }
}
.collapsible-menu__item {
  display: block;
  padding-left: 16px;

  &--disabled {
    color: #888;
  }
  a {
    color: ${colors.textMediumGrey};

    &:hover {
      color: ${colors.textDarkGrey};
    }
  }
}
`


// Usage: either set the 'children' or the 'options' property
//        if both are set children takes precedent
//        In options dict, either use onClick and target for the links
//
//        If titleAnchor is defined, a HashLink is displayed using that anchor
//        If not, title is used

const exampleOptions = [
  {
    text: 'action1',
    isDisabled: false,
    onClick: () => console.log('action1 click'),
  },
  {
    text: 'action2',
    target: '/some/url',
    isDisabled: false,
  },
  {
    text: 'action3',
    isDisabled: false,
  },
  {
    text: 'action4',
    isDisabled: false,
  },
  {
    text: 'action5',
    isDisabled: false,
  },
]

const Item = ({ text, isDisabled, target, onClick, entityType }) => {
  const classes = classNames({
    'collapsible-menu__item--disabled': isDisabled,
  }, 'collapsible-menu__item')

  const handler = () => {
    if (!isDisabled && typeof onClick === 'function') onClick()
  }

  return (
    <div className={classes} onClick={handler}>
      { target ? <Link to={target} aria-label={`Click to view ${text} ${entityType}`}>{text}</Link> : <a aria-label={`Click to view ${text} ${entityType}`}>{text}</a> }
    </div>
  )
}


const CollapsibleMenu = ({ title, titleAnchor, children, options }) => {
  const menuOptions = options ? options : exampleOptions

  const generateList = () => { return menuOptions.map((e) => {
    return <Item {...e} key={e.text} onClick={e.onClick} entityType={e.entityType}/>
  })}

  const menuId = self.crypto.randomUUID()
  const collapseMenuId = 'collapseMenu'+menuId+title
  const collapseBodyId = 'collapseBody'+menuId+title

  return (
    <StyledMenu>
    <div className="accordion collapsible-menu" id={collapseMenuId}>
      {titleAnchor ? (
      <div className="accordion-header collapsible-header">
        <div style={{ display: 'flex' }}>
          <div style={{ flex: '0 0 16px', marginTop: '12px' }}>
            <Icon id='toggle-icon' icon='fa-angle-down' data-toggle="collapse" data-target={'#'+collapseBodyId} aria-controls={collapseBodyId} />
          </div>
          <div className='title'>
            <HashLink smooth to={titleAnchor} aria-label={`Navigates to ${title} component`}>{title}</HashLink>
          </div>
        </div>
      </div>
      ) : (
      <div className="accordion-header collapsible-header" data-toggle="collapse" data-target={'#'+collapseBodyId} aria-controls={collapseBodyId}>
        <Icon id='toggle-icon' icon='fa-angle-down' />
        <div className='title' aria-label={`Navigate to ${title} section`} >{title}</div>
      </div>
      )}
      <div id={collapseBodyId} className="accordion-collapse collapse in" aria-label="Collapsible menu heading" data-parent={'#'+collapseMenuId}>
        {children ? children : generateList()}
      </div>
    </div>
    </StyledMenu>
  )
}

Item.propTypes = {
  text: PropTypes.string,
  title: PropTypes.string,
  target: PropTypes.string,
  isDisabled: PropTypes.bool,
  onClick: PropTypes.func,
  entityType: PropTypes.string,
}

CollapsibleMenu.propTypes = {
  title: PropTypes.string,
  titleAnchor: PropTypes.string,
  options: PropTypes.array,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
}

export default CollapsibleMenu
