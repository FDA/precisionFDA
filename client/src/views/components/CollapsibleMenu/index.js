import React from 'react'
import classNames from 'classnames'
import PropTypes from 'prop-types'
import { Link } from 'react-router-dom'
import { HashLink } from 'react-router-hash-link'
import { v4 as uuidv4 } from 'uuid'

import Icon from '../Icon'
import './style.sass'


// Usage: either set the 'children' or the 'options' property
//        if both are set children takes precedent
//        In options dict, either use onClick and target for the links
//
//        If titleAnchor is defined, a HashLink is displayed using that anchor
//        If not, title is used

const exampleOptions = [
  {
    text: 'action1',
    isDisabled: false,
    onClick: () => console.log('action1 click'),
  },
  {
    text: 'action2',
    target: '/some/url',
    isDisabled: false,
  },
  {
    text: 'action3',
    isDisabled: false,
  },
  {
    text: 'action4',
    isDisabled: false,
  },
  {
    text: 'action5',
    isDisabled: false,
  },
]

const Item = ({ text, isDisabled, target, onClick, entityType }) => {
  const classes = classNames({
    'collapsible-menu__item--disabled': isDisabled,
  }, 'collapsible-menu__item')

  const handler = () => {
    if (!isDisabled && typeof onClick === 'function') onClick()
  }

  return (
    <div className={classes} onClick={handler}>
      { target ? <Link to={target} aria-label={`Click to view ${text} ${entityType}`}>{text}</Link> : <a aria-label={`Click to view ${text} ${entityType}`}>{text}</a> }
    </div>
  )
}


const CollapsibleMenu = ({ title, titleAnchor, children, options }) => {
  const menuOptions = options ? options : exampleOptions

  const generateList = () => { return menuOptions.map((e) => {
    return <Item {...e} key={e.text} onClick={e.onClick} entityType={e.entityType}/>
  })}

  const menuId = uuidv4()
  const collapseMenuId = 'collapseMenu'+menuId+title
  const collapseBodyId = 'collapseBody'+menuId+title

  return (
    <div className="accordion collapsible-menu" id={collapseMenuId}>
      {titleAnchor ? (
      <div className="accordion-header collapsible-header">
        <div style={{ display: 'flex' }}>
          <div style={{ flex: '0 0 16px', marginTop: '12px' }}>
            <Icon id='toggle-icon' icon='fa-angle-down' data-toggle="collapse" data-target={'#'+collapseBodyId} aria-controls={collapseBodyId} />
          </div>
          <div className='title'>
            <HashLink smooth to={titleAnchor} aria-label={`Navigates to ${title} component`}>{title}</HashLink>
          </div>
        </div>
      </div>
      ) : (
      <div className="accordion-header collapsible-header" data-toggle="collapse" data-target={'#'+collapseBodyId} aria-controls={collapseBodyId}>
        <Icon id='toggle-icon' icon='fa-angle-down' />
        <div className='title' aria-label={`Navigate to ${title} section`} >{title}</div>
      </div>
      )}
      <div id={collapseBodyId} className="accordion-collapse collapse in" aria-label="Collapsible menu heading" data-parent={'#'+collapseMenuId}>
        {children ? children : generateList()}
      </div>
    </div>
  )
}

Item.propTypes = {
  text: PropTypes.string,
  title: PropTypes.string,
  target: PropTypes.string,
  isDisabled: PropTypes.bool,
  onClick: PropTypes.func,
  entityType: PropTypes.string,
}

CollapsibleMenu.propTypes = {
  title: PropTypes.string,
  titleAnchor: PropTypes.string,
  options: PropTypes.array,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
}

export default CollapsibleMenu
