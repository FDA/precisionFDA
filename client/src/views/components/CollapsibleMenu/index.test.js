import React from 'react'
import { shallow } from 'enzyme'

import CollapsibleMenu from './index'


describe('CollapsibleMenu', () => {
    it.skip('matches snapshot', () => {
      const wrapper = shallow(<CollapsibleMenu />)
  
      expect(wrapper).toMatchSnapshot()
    })
})