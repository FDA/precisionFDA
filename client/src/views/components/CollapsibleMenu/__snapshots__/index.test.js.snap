// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleMenu matches snapshot 1`] = `
<div
  className="accordion collapsible-menu"
  id="collapseMenu83cd99b9-51ea-48b8-a2ba-f4f01e335d08"
>
  <div
    aria-controls="collapseBody83cd99b9-51ea-48b8-a2ba-f4f01e335d08"
    aria-expanded="true"
    className="collapsible-header collapsed"
    data-target="#collapseBody83cd99b9-51ea-48b8-a2ba-f4f01e335d08"
    data-toggle="collapse"
  >
    <Icon
      icon="fa-angle-down"
      id="toggle-icon"
    />
    <div
      className="title"
    />
    Â 
  </div>
  <div
    aria-labelledby="collapsibleMenuHeading"
    className="collapse"
    data-parent="#collapseMenu83cd99b9-51ea-48b8-a2ba-f4f01e335d08"
    id="collapseBody83cd99b9-51ea-48b8-a2ba-f4f01e335d08"
  >
    <Item
      isDisabled={false}
      key="action1"
      onClick={[Function]}
      text="action1"
    />
    <Item
      isDisabled={false}
      key="action2"
      text="action2"
    />
    <Item
      isDisabled={false}
      key="action3"
      text="action3"
    />
    <Item
      isDisabled={false}
      key="action4"
      text="action4"
    />
    <Item
      isDisabled={false}
      key="action5"
      text="action5"
    />
  </div>
</div>
`;
